File: pytorch/03_train_model/inference.py
import torch
import torchvision
import time


class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.premodel = torch.nn.Sequential(
            torch.nn.Conv2d(1, 6, 5),
            torch.nn.ReLU(),
            torch.nn.MaxPool2d(2),
            torch.nn.Conv2d(6, 16, 5),
            torch.nn.ReLU(),
            torch.nn.MaxPool2d(2),
            torch.nn.Flatten(),
            torch.nn.Linear(256, 120),
            torch.nn.ReLU(),
            torch.nn.Linear(120, 84),
            torch.nn.ReLU(),
            torch.nn.Linear(84, 10)
        )

    def forward(self, x):
        return self.premodel(x)


transform = torchvision.transforms.Compose([
    torchvision.transforms.ToTensor(),
    torchvision.transforms.Normalize([0.5], [0.5])
])

testset = torchvision.datasets.MNIST(
    root='../data', train=False, download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=64, shuffle=False)

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
net = torch.jit.load('./checkpoint/model.pt').to(device)
loss_fn = torch.nn.CrossEntropyLoss()

net.eval()
test_loss, test_acc = 0, 0

start = time.time()
with torch.no_grad():
    for images, labels in testloader:
        images = images.to(device=device)
        labels = labels.to(device=device, dtype=torch.long)

        output = net(images)
        loss = loss_fn(output, labels)

        test_loss += loss.item()
        test_acc += (labels == output.argmax(dim=1)).sum().item()
test_loss /= len(testloader)
test_acc /= len(testloader.dataset)

end = time.time()
print('test_loss: {:.4f} test_acc: {:.4f} time: {:.4f}'.format(
    test_loss, test_acc, end-start))

File: pytorch/03_train_model/inference.cpp
#include <torch/torch.h>

#include <chrono>
#include <iostream>

int main() {
    torch::Device device(torch::cuda::is_available() ? torch::kCUDA : torch::kCPU);
    std::cout << device << std::endl;

    // Define model
    torch::nn::Sequential net(
        torch::nn::Conv2d(1, 6, 5),
        torch::nn::ReLU(),
        torch::nn::MaxPool2d(2),
        torch::nn::Conv2d(6, 16, 5),
        torch::nn::ReLU(),
        torch::nn::MaxPool2d(2),
        torch::nn::Flatten(),
        torch::nn::Linear(256, 120),
        torch::nn::ReLU(),
        torch::nn::Linear(120, 84),
        torch::nn::ReLU(),
        torch::nn::Linear(84, 10));
    torch::load(net, "./checkpoint/model.pt");
    net->to(device);

    auto testset = torch::data::datasets::MNIST(
                       "../data/MNIST/raw",
                       torch::data::datasets::MNIST::Mode::kTest)
                       .map(torch::data::transforms::Normalize<>(0.5, 0.5))
                       .map(torch::data::transforms::Stack<>());
    auto testloader = torch::data::make_data_loader<
        torch::data::samplers::SequentialSampler>(
        std::move(testset),
        torch::data::DataLoaderOptions().batch_size(64).workers(2));

    float test_loss = 0, test_acc = 0;
    int count_batch = 0, count_sample = 0;

    net->eval();
    torch::NoGradGuard no_grad;
    
    auto start = std::chrono::high_resolution_clock::now();
    for (torch::data::Example<>& batch : *testloader) {
        torch::Tensor images = batch.data.to(device);
        torch::Tensor labels = batch.target.to(device);

        torch::Tensor output = net->forward(images);
        torch::Tensor loss = torch::nn::functional::cross_entropy(output, labels);
        test_acc += torch::argmax(output, 1).eq(labels).sum().item<int64_t>();
        test_loss += loss.item<float>();

        count_batch += 1;
        count_sample += batch.data.size(0);
    }
    test_acc /= count_sample;
    test_loss /= count_batch;

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end - start;

    printf("test_loss: %.4f test_acc: %.4f time: %0.4f\n",
           test_loss, test_acc, diff.count());
    return 0;
}